packages/http/src/validator/validators/params.ts

```diff
@@ -15,45 +16,42 @@ export class HttpParamsValidator<Target, Spec extends IHttpParam> implements IHt
 
   public validate(target: Target, specs: Spec[]): IPrismDiagnostic[] {
     const { _registry: registry, _prefix: prefix, _style: style } = this;
-    return specs.reduce<IPrismDiagnostic[]>((results, spec) => {
-      if (!(spec.name in target) && spec.required === true) {
-        results.push({
-          path: [prefix, spec.name],
-          code: 'required',
-          message: `Missing ${spec.name} ${prefix} param`,
-          severity: DiagnosticSeverity.Error,
-        });
-
-        // stop further checks
-        return results;
-      }
-
-      const resolvedStyle = spec.style || style;
-      if (spec.content && spec.content.schema && target[spec.name]) {
-        const deserializer = registry.get(resolvedStyle);
 
-        if (deserializer) {
-          Array.prototype.push.apply(
-            results,
-            validateAgainstSchema(
-              deserializer.deserialize(spec.name, target, spec.content.schema, spec.explode || false),
-              spec.content.schema,
-              prefix,
-            ),
+    const deprecatedWarnings = specs.filter(spec => spec.deprecated).map(spec => ({
+      path: [prefix, spec.name],
+      code: 'deprecated',
+      message: `${upperFirst(prefix)} param ${spec.name} is deprecated`,
+      severity: DiagnosticSeverity.Warning,
+    }));
+
+    const schema = createJsonSchemaFromParams(specs);
+
+    const parameterValues = pickBy(
+      mapValues(keyBy(specs, s => s.name.toLowerCase()), el => {
+        const resolvedStyle = el.style || style;
+        const deserializer = registry.get(resolvedStyle);
+        if (deserializer)
+          return deserializer.deserialize(
+            el.name.toLowerCase(),
+            target,
+            schema.properties && (schema.properties[el.name] as JSONSchema4),
+            el.explode || false,
           );
-        }
-      }
-
-      if (spec.deprecated === true) {
-        results.push({
-          path: [prefix, spec.name],
-          code: 'deprecated',
-          message: `${upperFirst(prefix)} param ${spec.name} is deprecated`,
-          severity: DiagnosticSeverity.Warning,
-        });
-      }
-
-      return results;
-    }, []);
+
+        return undefined;
+      }),
+    );
+
+    return validateAgainstSchema(parameterValues, schema, prefix).concat(deprecatedWarnings);
   }
 }
+
+function createJsonSchemaFromParams(params: IHttpParam[]): JSONSchema4 | JSONSchema6 | JSONSchema7 {
+  const schema: JSONSchema4 | JSONSchema6 | JSONSchema7 = {
+    type: 'object',
+    properties: pickBy(mapValues(keyBy(params, p => p.name.toLowerCase()), 'schema')) as JSONSchema4,
```

Comment:
The more I look at the `x.toLowerCase` the more I come to a conclusion that this should be done at a class level.
At some point someone will forget to lower case a value and we'll get an error. If, instead, we used a class that would override setters (and/or getters) and lowercase everything it would be much simpler to write this code w/o risking hard to find errors. 
I don't think it's worth refactoring now, but WDYT as a general rule?

Reply:
I like the idea of performing this automagically — but I do not really follow the example practically (through a class) — can you clarify/make a minimal example?

I was thinking of using [caseless](https://www.npmjs.com/package/caseless) again, which we're already employing in the codebase.

Reply:
So we currently mostly use interfaces only (which is awesome!) and we don't use classes.
Imagine that you pass a param `params: IHttpParam[]` but the instance of it is in fact `params: CaselessHttpParam[]` (or whatever the name, could be simply `HttpParam`

Then the implementation would simply be:
```ts
class HttpParam implements IHttpParam {
   // I made that decorator up but it would effectively implement a setter and a getter and lowercase stuff.
   @caseless
   public name: String;
}
```

or if possible

```ts
@caseless
class HttpParam implements IHttpParam {}
```

Reply:
Then in code you'd simply do 
`keyBy(params, p => p.name)`
In fact, you wouldn't even have to because all would be lowercase in the first place and there would be no dups.

Reply:
Ok clear, let me see if I can come up with something!

Reply:
I'll have to pass on on this one unfortunately — it's really too much long to do know, I'll come up with something better when we will revisit the deserialisers/validators part.